lib/rust/ensogl/core/src/debug/monitor.rs

```diff
@@ -931,3 +884,113 @@ stats_sampler!(
     0,
     1.0
 );
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use assert_approx_eq::assert_approx_eq;
+
+
+    #[test]
+    fn frame_time() {
+        // Note: 60 FPS means there's 16.6(6) ms budget for 1 frame. The test will be written under
+        // assumption we're trying to be around this FPS.
+
+        let stats: Stats = default();
+        let sampler: Box<dyn Sampler> = Box::new(FrameTime::new(&stats));
+        let mut t = 0.0f64;
+
+        // Frame 1: simulate we managed to complete the work in 10ms, and then we wait 6ms before
+        // starting next frame.
+        stats.begin_frame(t);
+        t += 10.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 10.0);
+        assert!(matches!(sampler.check(), ValueCheck::Correct));
+        t += 6.0;
+
+        // Frame 2: simulate we managed to complete the work in 5ms, and then we wait 11ms before
+        // starting next frame.
+        stats.begin_frame(t);
+        t += 5.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 5.0);
+        assert!(matches!(sampler.check(), ValueCheck::Correct));
+        t += 11.0;
+
+        // Frame 3: simulate we went over the budget of 16.6(6) ms, at 30ms. No extra delay
+        // afterwards before starting next frame.
+        stats.begin_frame(t);
+        t += 30.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 30.0);
+        assert!(matches!(sampler.check(), ValueCheck::Warning));
+
+        // Frame 4: simulate a really slow frame (1000ms), crossing the configured error threshold.
+        stats.begin_frame(t);
+        t += 1000.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 1000.0);
+        assert!(matches!(sampler.check(), ValueCheck::Error));
+    }
+
+    #[test]
+    fn fps() {
+        // Note: 60 FPS means there's 16.6(6) ms budget for 1 frame. The test will be written under
+        // assumption we're trying to be around this FPS.
+
+        let stats: Stats = default();
+        let sampler: Box<dyn Sampler> = Box::new(Fps::new(&stats));
+        // Note: we can't use t=0.0 here, as this would throw off inner logic of Fps.
+        let mut t = 123.0f64;
```

Comment:
This not is not understandable by the reader. What does it mean that it "would throw off inner logic of Fps.". Reading that I feel that its very bad that we can't have a test with time = 0.

Reply:
I worked on improving the comment; would you find the new phrasing good enough?